rustls/src/msgs/codec.rs

```diff
@@ -174,88 +177,149 @@ impl Codec for u64 {
     }
 }
 
-pub fn encode_vec_u8<T: Codec>(bytes: &mut Vec<u8>, items: &[T]) {
-    let len_offset = bytes.len();
-    bytes.push(0);
+/// Represent a Rust `Vec` together with the length prefix type used to encode it in TLS.
+///
+/// This implements various traits to make it easy to use in place of a normal `Vec` and to
+/// easily convert between `Vec` and `TlsVec` at API boundaries.
+#[derive(Clone, Debug)]
+pub struct TlsVec<L, T> {
+    items: Vec<T>,
+    len: PhantomData<L>,
+}
+
+impl<L: LengthPrefix, T: Codec> Codec for TlsVec<L, T>
+where
+    usize: From<L>,
+{
```

Comment:
In order to avoid code bloat due to monomorphizing, as much of the code as possible in these implementations should be factoed out into non-generic functions.